File: src/main/java/dev/klomptech/jbark/parser/share/ShareParser.java
package dev.klomptech.jbark.parser.share;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.Parser;
import java.util.ArrayList;
import java.util.List;

public class ShareParser {

  private final Parser parser;

  public ShareParser(final Parser parser) {
    this.parser = parser;
  }

  // Pass one toy from dog to a single other dog (like: she passes a toy to the beagle)
  public AstNode parsePassLine() throws BarkError {
    int line = parser.peek().line();
    int passAt = findWordAfter(-1, Keywords.PASSES_STATEMENT_KEYWORDS);
    if (passAt < 0) {
      return null;
    }
    int toAt = findWordAfter(passAt, Keywords.TO_WORDS);
    if (toAt < 0) {
      return null;
    }
    String giver = findDogBefore(passAt);
    if (giver == null) {
      throw new BarkError(line, "Only dogs can pass toys, name who is giving first");
    }
    List<String> recipients = findDogsAfter(toAt + 1);
    if (recipients.isEmpty()) {
      throw new BarkError(line, "Passing needs one friend, give the name of who gets the item");
    }
    if (recipients.size() > 1) {
      throw new BarkError(line, "One pass, one friend. Use shares when more dogs get a slice");
    }
    parser.consumeUntilLineEnd();
    return new AstNode.Pass(giver, recipients.get(0), line);
  }

  // Split a stash with other dogs (e.g. labrador shares her cookie jar with the beagle)
  public AstNode parseShareLine() throws BarkError {
    int line = parser.peek().line();
    int sharesAt = findWordAfter(-1, Keywords.SHARES_STATEMENT_KEYWORDS);
    if (sharesAt < 0) {
      return null;
    }
    int withAt = findWordAfter(sharesAt, Keywords.WITH_WORDS);
    if (withAt < 0) {
      return null;
    }
    String giver = findDogBefore(sharesAt);
    if (giver == null) {
      throw new BarkError(line, "Only dogs can share. Name who is giving first");
    }
    String stash = null;
    if (withAt > sharesAt + 1) {
      int lineStart = parser.cursorIndex();
      stash =
          ConfigLoader.resolveCollectionFromTokens(
              parser.allTokens(), lineStart + sharesAt + 1, lineStart + withAt);
    }
    List<String> recipients = findDogsAfter(withAt + 1);
    if (recipients.isEmpty()) {
      throw new BarkError(
          line, "The dogs need a friend to share with. Name another dog after with");
    }
    parser.consumeUntilLineEnd();
    return new AstNode.Share(giver, stash, recipients, line);
  }

  private int findWordAfter(final int afterOffset, final List<String> words) {
    int start = afterOffset + 1;
    for (int offset = start; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()))) {
        return offset;
      }
    }
    return -1;
  }

  private String findDogBefore(final int beforeOffset) {
    for (int offset = 0; offset < beforeOffset; offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.isDogSubjectReference(word, token.value())) {
        return word;
      }
    }
    return null;
  }

  private List<String> findDogsAfter(final int startOffset) throws BarkError {
    int line = parser.peek().line();
    int lineStart = parser.cursorIndex();
    int lineEnd = parser.lineEndIndex();
    List<String> dogs = new ArrayList<>();
    int offset = startOffset;
    while (offset < lineEnd - lineStart) {
      int abs = lineStart + offset;
      Token token = parser.allTokens().get(abs);
      if (token.is(TokenType.COMMA)) {
        offset++;
        continue;
      }
      if (!token.is(TokenType.IDENTIFIER)) {
        offset++;
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.THE_WORDS.contains(word)
          || Keywords.ARTICLE_WORDS.contains(word)
          || Keywords.LIST_AND_WORDS.contains(word)
          || Keywords.WITH_WORDS.contains(word)
          || Keywords.SHARES_STATEMENT_KEYWORDS.contains(word)) {
        offset++;
        continue;
      }
      String breed = ConfigLoader.resolveNameFromTokens(parser.allTokens(), abs, lineEnd);
      if (breed != null
          && (ConfigLoader.isBreed(breed)
              || Keywords.isBreedPronounWord(breed)
              || Keywords.isPetNameWord(breed, token.value()))) {
        dogs.add(breed);
        offset = advancePastNamePhrase(offset, breed);
        continue;
      }
      if (ConfigLoader.isBreed(word)
          || Keywords.isBreedPronounWord(word)
          || Keywords.isPetNameWord(word, token.value())) {
        dogs.add(word);
        offset++;
        continue;
      }
      if (!Keywords.isIgnored(word)) {
        throw new BarkError(line, "\"" + word + "\" is not a dog the pack can share with.");
      }
      offset++;
    }
    return dogs;
  }

  private int advancePastNamePhrase(final int offsetFromLineStart, final String name) {
    int lineStart = parser.cursorIndex();
    int lineEnd = parser.lineEndIndex();
    List<String> words = new ArrayList<>();
    List<Integer> positions = new ArrayList<>();
    for (int abs = lineStart + offsetFromLineStart; abs < lineEnd; abs++) {
      Token token = parser.allTokens().get(abs);
      if (token.is(TokenType.IDENTIFIER)) {
        words.add(parser.normalise(token.value()));
        positions.add(abs);
      }
    }
    for (int length = Math.min(4, words.size()); length >= 1; length--) {
      for (int start = 0; start <= words.size() - length; start++) {
        String phrase = String.join(" ", words.subList(start, start + length));
        if (name.equals(ConfigLoader.resolveNamePhrase(phrase))) {
          return positions.get(start + length - 1) + 1 - lineStart;
        }
      }
    }
    return offsetFromLineStart + 1;
  }
}

File: pbark/pbark/parser/share/share_parser.py
from __future__ import annotations

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords


class ShareParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_pass_line(self):
        line = self._parser.peek().line
        pass_at = self._find_word_after(-1, Keywords.PASSES_STATEMENT_KEYWORDS)
        if pass_at < 0:
            return None
        to_at = self._find_word_after(pass_at, Keywords.TO_WORDS)
        if to_at < 0:
            return None
        giver = self._find_dog_before(pass_at)
        if giver is None:
            raise BarkError(line, "Only dogs can pass toys, name who is giving first")
        recipients = self._find_dogs_after(to_at + 1)
        if not recipients:
            raise BarkError(line, "Passing needs one friend, give the name of who gets the item")
        if len(recipients) > 1:
            raise BarkError(line, "One pass, one friend. Use shares when more dogs get a slice")
        self._parser.consume_until_line_end()
        return AstNode.Pass(giver, recipients[0], line)

    def parse_share_line(self):
        line = self._parser.peek().line
        shares_at = self._find_word_after(-1, Keywords.SHARES_STATEMENT_KEYWORDS)
        if shares_at < 0:
            return None
        with_at = self._find_word_after(shares_at, Keywords.WITH_WORDS)
        if with_at < 0:
            return None
        giver = self._find_dog_before(shares_at)
        if giver is None:
            raise BarkError(line, "Only dogs can share. Name who is giving first")
        stash = None
        if with_at > shares_at + 1:
            line_start = self._parser.cursor_index()
            stash = ConfigLoader.resolve_collection_from_tokens(
                self._parser.all_tokens(), line_start + shares_at + 1, line_start + with_at
            )
        recipients = self._find_dogs_after(with_at + 1)
        if not recipients:
            raise BarkError(
                line, "The dogs need a friend to share with. Name another dog after with"
            )
        self._parser.consume_until_line_end()
        return AstNode.Share(giver, stash, recipients, line)

    def _find_word_after(self, after_offset: int, words: list[str]) -> int:
        start = after_offset + 1
        offset = start
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words:
                return offset
            offset += 1
        return -1

    def _find_dog_before(self, before_offset: int) -> str | None:
        for offset in range(before_offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                continue
            word = self._parser.normalise(token.value)
            if Keywords.is_dog_subject_reference(word, token.value):
                return word
        return None

    def _find_dogs_after(self, start_offset: int) -> list[str]:
        line = self._parser.peek().line
        line_start = self._parser.cursor_index()
        line_end = self._parser.line_end_index()
        dogs: list[str] = []
        offset = start_offset
        while offset < line_end - line_start:
            abs_idx = line_start + offset
            token = self._parser.all_tokens()[abs_idx]
            if token.is_(TokenType.COMMA):
                offset += 1
                continue
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if (
                word in Keywords.THE_WORDS
                or word in Keywords.ARTICLE_WORDS
                or word in Keywords.LIST_AND_WORDS
                or word in Keywords.WITH_WORDS
                or word in Keywords.SHARES_STATEMENT_KEYWORDS
            ):
                offset += 1
                continue
            breed = ConfigLoader.resolve_name_from_tokens(
                self._parser.all_tokens(), abs_idx, line_end
            )
            if breed is not None and (
                ConfigLoader.is_breed(breed)
                or Keywords.is_breed_pronoun_word(breed)
                or Keywords.is_pet_name_word(breed, token.value)
            ):
                dogs.append(breed)
                offset = self._advance_past_name_phrase(offset, breed)
                continue
            if (
                ConfigLoader.is_breed(word)
                or Keywords.is_breed_pronoun_word(word)
                or Keywords.is_pet_name_word(word, token.value)
            ):
                dogs.append(word)
                offset += 1
                continue
            if not Keywords.is_ignored(word):
                raise BarkError(line, f'"{word}" is not a dog the pack can share with.')
            offset += 1
        return dogs

    def _advance_past_name_phrase(self, offset_from_line_start: int, name: str) -> int:
        line_start = self._parser.cursor_index()
        line_end = self._parser.line_end_index()
        words: list[str] = []
        positions: list[int] = []
        for abs_idx in range(line_start + offset_from_line_start, line_end):
            token = self._parser.all_tokens()[abs_idx]
            if token.is_(TokenType.IDENTIFIER):
                words.append(self._parser.normalise(token.value))
                positions.append(abs_idx)
        for length in range(min(4, len(words)), 0, -1):
            for start in range(len(words) - length + 1):
                phrase = " ".join(words[start : start + length])
                if name == ConfigLoader.resolve_name_phrase(phrase):
                    return positions[start + length - 1] + 1 - line_start
        return offset_from_line_start + 1
